src/transformers/generation/flax_logits_process.py

```diff
@@ -455,3 +456,82 @@ def handle_cumulative_probs(logprobs_k, scores_k):
         scores = jax.vmap(handle_cumulative_probs)(logprobs, scores)
 
         return scores
+
+
+class FlaxNoRepeatNGramLogitsProcessor(FlaxLogitsProcessor):
+    r"""
+    [`FlaxLogitsProcessor`] that enforces no repetition of n-grams. See
+    [Fairseq](https://github.com/pytorch/fairseq/blob/a07cb6f40480928c9e0548b737aadd36ee66ac76/fairseq/sequence_generator.py#L345).
+
+    Args:
+        ngram_size (`int`):
+            All ngrams of size `ngram_size` can only occur once.
+    """
+
+    def __init__(self, ngram_size: int):
+        if not isinstance(ngram_size, int) or ngram_size <= 0:
+            raise ValueError(f"`ngram_size` has to be a strictly positive integer, but is {ngram_size}")
+        self.ngram_size = ngram_size
+
+    def get_previous_ngrams(self, input_ids: jnp.ndarray, vocab_size: int, cur_len: int):
+        """
+        get a matrix of size (batch_size,) + (vocab_size,)*n (for n-grams) that
+        represent the n-grams that occured previously.
+        The BCOO representation allow to store only the few non-zero entries, instead of the full (huge) matrix
+        """
+        batch_size, seq_len = input_ids.shape
+
+        all_update_indices = jnp.array(
+            [
+                [
+                    b,
+                ]
+                + [input_ids[b, i + j] for j in range(self.ngram_size)]
+                for i in range(seq_len - (self.ngram_size - 1))
+                for b in range(batch_size)
+            ]
+        )
+
+        data = jnp.ones((all_update_indices.shape[0],), dtype=jnp.uint16)
+        data = data * (jnp.arange(data.shape[0]) < batch_size * (cur_len - (self.ngram_size - 1)))  # ignore the n-grams not yet generated
```

Comment:
from my experience with jax, in order for the code to work with jit, you cannot use arrays of shapes that are not fixed. This is why i opted to pad the indices to have a known size. There may be more efficient ways to do it but I found that this is working, as opposed to `[]` slices, or even [dynamic slices](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dynamic_slice.html), because `cur_length` changes.

Reply:
I see :) I made the suggestion based on my past XLA+TF experience -- slicing `input_ids = input_ids[:, :cur_len]` is allowed there ([example](https://github.com/huggingface/transformers/blob/8dd4ce6f2cea1316cb5bd5ccbd348310ced17856/src/transformers/generation/tf_logits_process.py#L281))

But then again, JAX is usually stricter. Let's keep as you suggested 🤗

Reply:
i think it also works in jax in some situations, but in this case the error when using jit is kind of explicit

```python
IndexError: Array slice indices must have static start/stop/step to be used with NumPy indexing syntax. Found slice(None, Traced<ShapedArray(int32[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>, None). To index a statically sized array at a dynamic position, try lax.dynamic_slice/dynamic_update_slice (JAX does not support dynamically sized arrays within JIT compiled functions).
```

anyway, i took some time to benchmark different ways of doing this kind of operation, and found that in this instance, using `jax.lax.fori_loop` to update the `all_update_indices` is significantly faster than using dynamic slices updates (with this + removed useless operations, i measured >10x speedup for the jitted function). there might still be room for improvement